pkg/server/endpoints/eventTracker_test.go

```diff
@@ -0,0 +1,1120 @@
+			pollDuration: time.Hour * time.Duration(1),
+
+			expectedPollPeriods: 60,
+		},
+		{
+			name:         "polling rounds up",
+			pollInterval: time.Minute * time.Duration(3),
+			pollDuration: time.Minute * time.Duration(10),
+
+			expectedPollPeriods: 4,
+		},
+	} {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			pollPeriods := endpoints.PollPeriods(tt.pollInterval, tt.pollDuration)
+
+			require.Equal(t, tt.expectedPollPeriods, pollPeriods, "interval %s, polled over %s yeilds %d poll periods, not %d poll periods", tt.pollInterval.String(), tt.pollDuration.String(), pollPeriods, tt.expectedPollPeriods)
+		})
+	}
+}
+
+func TestNewEventTracker(t *testing.T) {
+	for _, tt := range []struct {
+		name        string
+		pollPeriods uint
+		boundaries  []uint
+
+		expectedInitialPolls uint
+		expectedPollPeriods  uint
+		expectedPolls        uint
+		expectedBoundaries   []uint
+	}{
+		{
+			name:        "polling always polls at least once",
+			pollPeriods: 0,
+			boundaries:  []uint{},
+
+			expectedInitialPolls: 1,
+			expectedPollPeriods:  1,
+			expectedPolls:        1,
+			expectedBoundaries:   []uint{},
+		},
+		{
+			name:        "polling once, pre-boundary",
+			pollPeriods: 1,
+			boundaries:  []uint{},
+
+			expectedInitialPolls: 1,
+			expectedPollPeriods:  1,
+			expectedPolls:        1,
+			expectedBoundaries:   []uint{},
+		},
+		{
+			name:        "polling once, in one bucket boundary",
+			pollPeriods: 1,
+			boundaries:  []uint{0},
+
+			expectedInitialPolls: 0,
+			expectedPollPeriods:  1,
+			expectedPolls:        1,
+			expectedBoundaries:   []uint{0},
+		},
+		{
+			name:        "polling twice, both initial",
+			pollPeriods: 2,
+			boundaries:  []uint{},
+
+			expectedInitialPolls: 2,
+			expectedPollPeriods:  2,
+			expectedPolls:        2,
+			expectedBoundaries:   []uint{},
+		},
+		{
+			name:        "polling twice, once initial, once in one bucket boundary",
+			pollPeriods: 2,
+			boundaries:  []uint{1},
+
+			expectedInitialPolls: 1,
+			expectedPollPeriods:  2,
+			expectedPolls:        2,
+			expectedBoundaries:   []uint{1},
+		},
+		{
+			name:        "polling once, in two bucket boundary",
+			pollPeriods: 2,
+			boundaries:  []uint{0},
+
+			expectedInitialPolls: 0,
+			expectedPollPeriods:  2,
+			expectedPolls:        1,
+			expectedBoundaries:   []uint{0},
+		},
+		{
+			name:        "polling once, in three bucket boundary",
+			pollPeriods: 3,
+			boundaries:  []uint{0},
+
+			expectedInitialPolls: 0,
+			expectedPollPeriods:  3,
+			expectedPolls:        1,
+			expectedBoundaries:   []uint{0},
+		},
+		{
+			name:        "polling six times in exponential backoff",
+			pollPeriods: 120,
+			boundaries:  []uint{0, 2, 6, 14, 30, 62},
+
+			expectedInitialPolls: 0,
+			expectedPollPeriods:  120,
+			expectedPolls:        6,
+			expectedBoundaries:   []uint{0, 2, 6, 14, 30, 62},
+		},
+		{
+			name:        "distributed linear polling for a while, then exponential",
```

Comment:
so iirc, this could help in testing if DB client's increment strategy changes during runtime? or are the boundaries not related to the the increments at all?

Reply:
Every time you check to see if something is to be schedule, it's time increments.  A boundary is one region of time (for example, between the 6th check and the 14th check) when that item should be checked once.

The hash's remainder when divided by the size of the boundary will determine which time within that boundary the item will be polled.  And number divided can only have one remainder, so this ensure that every polled item gets polled just once in the boundary>

The main reason to make this configurable (by allowing someone to pass boundaries to the eventTracker) is because we don't know what the most effective polling approach will be.  It might even be different for different deployments / organizational policies.  The `BoundaryBuilder` is just a default first-attempt that's known to be better than "poll every time".

When talking to the people impacted the most by event auto increments greater than 1, they didn't have a lot of confidence that they could predict the correct approach; that's why it is configurable.  If this isn't sufficient, we will expose the boundary selection in some way through the configuration file.

Reply:
The strategy doesn't change during runtime, but it's not hard coded.  That permits one to alter it in the future for different runtimes.

Right now there's no idea of what kinds of strategies Uber might need, but it's clearly not exponential (the default strategy people mention when they hear backoff).  It might be linear up to a "minimum polling interval" time, it might be something else.  When Uber gets this in their environment (800,000 to 1,600,000 tracked entries estimated) we can determine what needs to be adjusted in future PRs.

Reply:
@amoore877 Sorry, I just reread this and discovered that I was thinking of the program's strategy changing, not the database's strategy.

Yes, if the database's strategy changes at runtime, the program will automatically adjust.  It will just be different "patterns" of skipped items, if any are skipped at all.  That's intentional, because the database setup that gives the current algorithm so much trouble isn't a "skip every other" pattern, it is a "one database writes even ids, the other writes odd ids"  That prevents the two databases from having to synchronize on ids or communicate when an id is issued.  

As we never know when the write database will switch, and there's no guarantee that one db will fully finish its work before a switch, just figuring out the ones to skip (a previous suggestion) doesn't really work.  Additionally, there is the ability to set up 3 or more databases, such that each one uses ids that don't conflict with the others under this kind of setup pattern.

That's the reason there are unit tests that check the items distribute evenly when they are incrementing by 2 (even), by 2 (odd), by 3 (divides evenly), by 3 (divides with remainder of 1), by 3 (divides with remainder of 2) and so on.  Some programming flaws can make it such that an item that always divides by 2 evenly would only use 1/2 of the time slots, an item that always divides by 3 evenly would only use 1/3 of the time slots, etc.

Reply:
As far as the boundaries not being related to the increments, they aren't.  But no matter what the increment is, any item being polled by any increment should be polled only once per boundary.  Also, all elements incrementing by whatever amount they are incrementing by should be polled evenly across any sized boundary.

Unit testing only tests for increments between 2 and 5 (at every divisibility remainder) and boundaries between 2 and 5 (gotta cut the tests off at some time).  If these are found to be "not enough" it's always possible to add more using the table-driven approach that exists already.

Now each boundary can be different, so you basically implement the algorithm to back off by assembling lots of boundaries.  The change in polling with respect to time (polls/time) is set by the width of the boundary.  If they are all equal, it's linear polling, if they are all offset by an exponential difference, it's exponential.  If you mix in different boundary widths, you can create nearly any polling algorithm one wishes (exponential resetting at every hour of polling, any algorithm capped by a certain minimum polling, linera at one rate that then decreased to another, etc), but this does push the complexity of the algorithm;s implementation into the boundary choices.